README.md
# PyFiNeR: Fitting Near-infrared RR Lyrae light curves

This routine implements the RR Lyrae near-infrared light curve fitting techniques described by Hajdu et al. (2018).

The K-band light curve are fit with the Fourier transformed principal components coefficients;
the H-band data is fit with the same light-curve shape; the shape of the J-band light curve is
approximated with the K-band principal component amplitudes.

This routine was developed for:
 - `Python` 2.7
 - `Numpy` 1.12
 - `Scipy` 1.0
 - `Matplotlib` 2.1

## Installation

Copy all files from the `bin` directory to the same directory in the system PATH.

## Usage

The routine requires six arguments to run; in order, these are:
- COMMENT: a string that is passed printed out in the end; usually the name of the variable;
can use quotation marks to carry additional information
- PERIOD: the initial period estimate of the variable
- FILE_K_LC: path of the K light curve file; must exist
- FILE_J_LC: path of the J light curve file; must exist, but can be an empty file
- FILE_H_LC: path of the H light curve file; must exist, but can be an empty file
- OUTPUT.pdf: name of the output pdf file

Given these arguments, the routine does the fit and gives back the following output:
- COMMENT: the string from the argument list
- PERIOD: the final period 
- K_MAG: magnitude averaged mean K magnitude
- K_INT: intensity averaged mean K magnitude
- J_MAG_MEAN : magnitude averaged mean J magnitude, calculated with the mean offset
- J_INT_MEAN : intensity averaged mean J magnitude, calculated with the mean offset
- J_MAG_MEDI : magnitude averaged mean J magnitude, calculated with the median offset
- J_INT_MEDI : intensity averaged mean J magnitude, calculated with the median offset
- H_MAG_MEAN : magnitude averaged mean H magnitude, calculated with the mean offset
- H_INT_MEAN : intensity averaged mean H magnitude, calculated with the mean offset
- H_MAG_MEDI : magnitude averaged mean H magnitude, calculated with the median offset
- H_INT_MEDI : intensity averaged mean H magnitude, calculated with the median offset
- U1: amplitude of the first principal component of the K light curve
- U2: amplitude of the second principal component of the K light curve
- U3: amplitude of the third principal component of the K light curve
- U4: amplitude of the fourth principal component of the K light curve
- RMSE: the Root Mean Square Error of the K-band fit
- COST: the final value of the Huber cost function
- COSTN: the Uber cost function per K-band light curve cost
- N_FINAL: the number of K-band light curve points in the final fit
- N_INITIAL: the initial number of K-band light curve points



CHANGES.txt
v1.0, 27.02.2018 -- Initial release.


test/test_pyfiner.sh
# A simple shell script to test the PyFiNeR fitting

echo "The output should be:"
echo "4779 2  0.5911760 15.166 15.164 16.472 16.468 16.469 16.465 15.579 15.577 15.574 15.572  0.698841  0.066634 -0.019603 -0.059386 0.0323 0.0544 0.000504  108  108"
echo " , while the 4779_2.pdf should look the same as the included 4779_2_comp.pdf"

pyfiner "4779 2" 0.591176 4779.k2 4779.j2 4779.h2 4779_2.pdf


pyfiner/test/test_pyfiner.sh
# A simple shell script to test the PyFiNeR fitting

# The output should be:
# 4779 2  0.5911760 15.166 15.164 16.472 16.468 16.469 16.465 15.579 15.577 15.574 15.572  0.698841  0.066634 -0.019603 -0.059386 0.0323 0.0544 0.000504  108  108
# , while the 4779_2.pdf should look the same as the included 4779_2_comp.pdf

python ../pyfiner.py "4779 2" 0.591176 4779.k2 4779.j2 4779.h2 4779_2.pdf


lcutil_pyfiner.py
# Light curve utilities

import numpy as np
from os import path

# We read in the necessary constants contained in the coefs_pca.npz file
# This file contains the harmonic coefficients of the Fourier transform
# of the K light curve PCA eignevectors, the corresponding harmonic
# coefficients of the J-band fits, as well as the upper and lower limits
# of the principal component amplitudes, in the format of U1, U2/U1, U3/U1
# and U4/U1

try:
    npzfile = np.load(path.dirname(__file__)+'/coefs_pca.npz')
    harmonic_coef_k = npzfile['k']
    harmonic_coef_j = npzfile['j']
    maxs            = npzfile['maxs']
    mins            = npzfile['mins']
except:
    print "Coefficient file not found"
    exit()

# The file we have read in has the information up to the 12 Fourier order
harmonic_order=12

# Simple function to calculate the phases of the timings of a light curve

def return_LC_phases(period, LC, epoch=0.0):
    return np.modf( (LC-epoch)/period )[0]


# Function to give back the sum of a Fourier series at given positions
# Can be used to give back the light-curve shape, if called with period=1.
# and positions (epochs) between 0 and 1

def return_harmonic_LC(order, period, coefs, intercept, positions):
    result=np.zeros(positions.size)
    for i in range(1,order+1):
        result=result +\
               coefs[i*2-2] *np.sin( i*2*np.pi*positions/period ) +\
               coefs[i*2-1] *np.cos( i*2*np.pi*positions/period)
    return result+intercept


# This function returns the residuals for a given light curve with
# amplitudes U1..U4
# Calls return_harmonic_LC, but passes the collapsed U*harmonic_coef
# to it
# This function is passed to the least_squares function of scipy.optimize
# as its first argument, which optimizes U1..U4 for the selected loss
# function (in our case, the Huber loss function)

def return_residuals(x, t, y):
    phases = return_LC_phases(x[2],t,x[1])
    U = np.array((x[3], x[3]*x[4], x[3]*x[5], x[3]*x[6]))
    return y - return_harmonic_LC(harmonic_order,x[2],(U*harmonic_coef_k.T).T.sum(axis=0),x[0],phases*x[2])


# Calculates the intensity averaged magnitude of a light curve
# The phases of the light curve have to be sampled evenly, but
# in a [0..1) way; should be also dense enough

def calc_int_average(lc):
    return -2.5*np.log10( np.average( (10.**(-lc/2.5)) ) )



pyfiner.py
#!/usr/bin/env python2.7
# PyFiNeR: Fitting Near-infrared RR Lyrae light curves

# This routine implements the RR Lyrae near-infrared light curve fitting techniques
# described in Hajdu et al. (2018)

# The K-band light curve are fit with the Fourier transformed principal components
# coefficients; the H-band data is fit with the same light-curve shape; the shape
# of the J-band light curve is approximated with the K-band principal component
# amplitudes

# This routine was developed for Python 2.7, Numpy 1.12, Scipy 1.0 and Matplotlib 2.1

# We import the necessary packages and make sure that matplotlib is not in interactive
# mode

import numpy as np
import matplotlib as mpl
mpl.use('Agg')
import matplotlib.pylab as plt
from scipy.optimize import least_squares
from sys import argv, exit
from lcutil_pyfiner import *
plt.ioff()


# We try to read in the arguments and open the light curves in the different bands
# We also calculate a basic shift to the times of the data

try:
    NAME      = argv[1]
    period    = np.float64(argv[2])
    K_LC_path = argv[3]
    J_LC_path = argv[4]
    H_LC_path = argv[5]
    saveto    = argv[6]
except:
    print "Usage: python",argv[0],"COMMENT PERIOD FILE_K_LC FILE_J_LC FILE_H_LC OUTPUT.pdf"
    print "COMMENT: a string passed through to the output, usually the name of the star"
    print "FILE_K_LC: path of the K light curve file"
    print "FILE_J_LC: path of the J light curve file, can be an empty file"
    print "FILE_H_LC: path of the H light curve file, can be an empty file"
    print "OUTPUT.pdf: name of the file to plot the results of the fit"

    exit()

try:
    LC_K=np.loadtxt(K_LC_path, unpack=True)
    jd_min=[np.min(LC_K[0,0])]
except:
    print "{:s}: K light curve file not found!".format(NAME)
    exit()

try:
    LC_J=np.loadtxt(J_LC_path, unpack=True)
    if LC_J.size==0:
        no_J = True
    else:
        no_J = False
        jd_min.append(np.min(LC_J[0]))
        
except:
    print "{:s}: J light curve file not found!".format(NAME)
    exit()

try:
    LC_H=np.loadtxt(H_LC_path, unpack=True)
    if LC_H.size==0:
        no_H = True
    else:
        no_H = False
        jd_min.append(np.min(LC_H[0]))
except:
    print "{:s}: H light curve file not found!".format(NAME)
    exit()

shift = np.int(np.min(jd_min))-2
LC_K[0]  = LC_K[0] - shift


# We make a first cut in the light curves at +/- 0.5 mag around the median value,
# while also discarding every magnitude brighter than 9 and dimmer than 20

first_mask = (LC_K[1] > 9) * (LC_K[1]<20)
mask =   ( LC_K[1] < np.median(LC_K[1][first_mask]+0.5) )\
       * ( LC_K[1] > np.median(LC_K[1][first_mask]-0.5) )

LC_K2 = LC_K[:,mask]


# With VVV photometry, it is possible that we won't have enough points ot continue
# with the fitting after the first cut; in this case, we exit here

if (LC_K2.shape[1]<10):
    print NAME, "9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9 9", LC_K2.shape[1], LC_K.shape[1]
    exit()


# We start by defining the trusted regions
# We also define a few variables to store the costs and fits of the first round of fits

lower = np.array((-np.inf, 0.0, period-0.001, mins[0], mins[1], mins[2], mins[3]))
upper = np.array(( np.inf, 2.0, period+0.001, maxs[0], maxs[1], maxs[2], maxs[3]))

first_fits = np.zeros((5,7))
costs      = np.zeros(5)


# We fit the models starting from different phases, then choose the best one to
# be the starting point in the next step of the fitting; we also do a 3.5 sigma cut,
# where sigma is estimated from the Median Absolute Deviation

for i in xrange(5):
    parameters_to_fit =  np.asarray( (np.median(LC_K2[1]), 1.0+(-2+i)*period/5., period, 0.7, 0., 0., 0.))
    huber             = least_squares(return_residuals, x0=parameters_to_fit, args=(LC_K2[0], LC_K2[1]),
                                      loss='huber', f_scale=0.05, bounds=(lower, upper))
    first_fits[i] = huber.x
    costs[i]      = huber.cost

best_id       = np.argmin(costs)
residuals     = return_residuals(first_fits[best_id], LC_K2[0], LC_K2[1])
residuals_std = np.median(np.abs(residuals))*1.4826
mask_35sigma  = (np.abs(residuals) < 3.5*residuals_std )
LC_K3         = LC_K2[:,mask_35sigma]


# We refit the model on the remaining light-curve points, but starting from the phase
# determined in the previous step

parameters_to_fit = np.asarray( (np.median(LC_K3[1]), first_fits[best_id, 1], first_fits[best_id, 2],
                               first_fits[best_id, 3], first_fits[best_id, 4], first_fits[best_id, 5], first_fits[best_id, 6]))
huber = least_squares(return_residuals, x0=parameters_to_fit, args=(LC_K3[0], LC_K3[1]),
                      loss='huber', f_scale=0.05, bounds=(lower, upper))


# The vector U contains the fitted amplitudes of the individual principal components
U = np.array((huber.x[3], huber.x[3]*huber.x[4], huber.x[3]*huber.x[5], huber.x[3]*huber.x[6]))

residuals_K3_final = return_residuals(huber.x, LC_K3[0], LC_K3[1])
RMSE= np.sqrt( np.mean(residuals_K3_final**2))


# We start plotting the results
#

LC_K2_phases = return_LC_phases(huber.x[2], LC_K2[0], 0.0)
LC_K3_phases = LC_K2_phases[mask_35sigma]

plt.style.use('seaborn-dark-palette')

fig, ax1 = plt.subplots(3, sharex=True)
fig.set_size_inches(9,10)
fig.subplots_adjust(hspace=0.05)
ax1[0].invert_yaxis()
ax1[1].invert_yaxis()
ax1[2].invert_yaxis()

plt.xlim(-0.01,2.01)

centers_long = np.linspace(-0.01,2.01,202)
fitted_LC_final  = return_harmonic_LC(harmonic_order, 1.0, (U*harmonic_coef_k.T).T.sum(axis=0),
                                      huber.x[0], centers_long-huber.x[1]/huber.x[2])

K_int_mean   = calc_int_average(return_harmonic_LC(harmonic_order, 1.0, (U*harmonic_coef_k.T).T.sum(axis=0),
                                0.0, np.linspace(0.0,0.995,200)-huber.x[1]/huber.x[2]) + huber.x[0])

ax1[0].yaxis.set_major_locator(mpl.ticker.MultipleLocator(0.1))
ax1[1].yaxis.set_major_locator(mpl.ticker.MultipleLocator(0.1))
ax1[2].yaxis.set_major_locator(mpl.ticker.MultipleLocator(0.1))
ax1[0].set_ylabel('$\mathrm{K}_\mathrm{S}$')
ax1[1].set_ylabel('H')
ax1[2].set_ylabel('J')

ax1[0].plot(LC_K2_phases,   LC_K2[1],'r.')
ax1[0].plot(LC_K2_phases+1, LC_K2[1],'r.')
ax1[0].plot(LC_K3_phases,   LC_K3[1],'b.')
ax1[0].plot(LC_K3_phases+1, LC_K3[1],'b.')
ax1[0].plot(centers_long, fitted_LC_final, 'r-', lw=2)

# We only calculate the average magnitudes and plot the J and H light curves if they have data

if no_J == True:
    J_mag_mean   = 99.999
    J_mag_median = 99.999
    J_int_mean   = 99.999
    J_int_median = 99.999
else:
    LC_J[0] = LC_J[0] - shift
    LC_J_phases        = return_LC_phases(huber.x[2], LC_J[0], 0.0)
    J_predicted_shape  = return_harmonic_LC(harmonic_order, 1.0,        (U*harmonic_coef_j.T).T.sum(axis=0),
                                            0.0, centers_long-huber.x[1]/huber.x[2])
    
    J_predicted_at_obs = return_harmonic_LC(harmonic_order, huber.x[2], (U*harmonic_coef_j.T).T.sum(axis=0),
                                            0.0, LC_J[0]-huber.x[1])
    J_mag_mean   = np.mean(LC_J[1]-J_predicted_at_obs)
    J_mag_median = np.median(LC_J[1]-J_predicted_at_obs)
    J_int_mean   = calc_int_average(return_harmonic_LC(harmonic_order, 1.0, (U*harmonic_coef_j.T).T.sum(axis=0),
                                    0.0, np.linspace(0.0,0.995,200)-huber.x[1]/huber.x[2]) + J_mag_mean)
    J_int_median = J_int_mean + J_mag_median - J_mag_mean

    ax1[2].plot(centers_long, J_predicted_shape+J_mag_mean, 'k-', lw=2)
    ax1[2].plot(centers_long, J_predicted_shape+J_mag_median, 'k--', lw=2)
    ax1[2].plot(LC_J_phases,  LC_J[1],'b.', ms=15.0)
    ax1[2].plot(LC_J_phases+1,LC_J[1],'b.', ms=15.0)


if no_H == True:
    H_mag_mean   = 99.999
    H_mag_median = 99.999
    H_int_mean   = 99.999
    H_int_median = 99.999
else:
    LC_H[0] = LC_H[0] - shift
    LC_H_phases        = return_LC_phases(huber.x[2], LC_H[0], 0.0)
    
    H_predicted_at_obs = return_harmonic_LC(harmonic_order, huber.x[2], (U*harmonic_coef_k.T).T.sum(axis=0),
                                            0.0, LC_H[0]-huber.x[1])
    H_mag_mean   = np.mean(  LC_H[1]-H_predicted_at_obs)
    H_mag_median = np.median(LC_H[1]-H_predicted_at_obs)
    H_int_mean   = H_mag_mean   - huber.x[0] + K_int_mean
    H_int_median = H_mag_median - huber.x[0] + K_int_mean

    ax1[1].plot(centers_long, fitted_LC_final-huber.x[0]+H_mag_mean,   'k-',  lw=2)
    ax1[1].plot(centers_long, fitted_LC_final-huber.x[0]+H_mag_median, 'k--', lw=2)
    ax1[1].plot(LC_H_phases,  LC_H[1],'b.', ms=15.0)
    ax1[1].plot(LC_H_phases+1,LC_H[1],'b.', ms=15.0)


if LC_K3.shape[1] > 0:
    cost=huber.cost
else:
    cost=9.9999

textstr = '{:10s}          P: {:9.7f}  N: {:d} ({:d})\n'.format(NAME, huber.x[2], LC_K3.shape[1], LC_K.shape[1])
textstr+= 'RMSE: {:6.4f}   cost: {:6.4f}   cost/N: {:8.6f}\n'.format(RMSE, cost, cost/LC_K3.shape[1])
textstr+= 'K = {:6.3f}  J = {:6.3f} ({:6.3f})  H = {:6.3f} ({:6.3f}) mag\n'.format(huber.x[0], J_mag_mean, J_mag_median, H_mag_mean, H_mag_median)
textstr+= 'K = {:6.3f}  J = {:6.3f} ({:6.3f})  H = {:6.3f} ({:6.3f}) int'.format(K_int_mean, J_int_mean, J_int_median, H_int_mean, H_int_median)

props = dict(boxstyle='round', facecolor='wheat', alpha=0.5)
ax1[0].text(.02, 1.04,textstr,fontsize=11,
            horizontalalignment='left',verticalalignment='bottom',
            bbox=props,
            transform=ax1[0].transAxes)


# The figure is saved to a pdf with the name given as the 

plt.savefig(saveto, bbox_inches='tight', format='pdf')


# Finally, we print out the results of the fit

textstr = '{:7s} {:9.7f} {:6.3f} {:6.3f} {:6.3f} {:6.3f} '.format(NAME, huber.x[2], huber.x[0], K_int_mean, J_mag_mean, J_int_mean)
textstr+= '{:6.3f} {:6.3f} {:6.3f} {:6.3f} {:6.3f} {:6.3f} '.format(J_mag_median, J_int_median, H_mag_mean, H_int_mean, H_mag_median, H_int_median)
textstr+= '{:9.6f} {:9.6f} {:9.6f} {:9.6f} {:6.4f} {:6.4f} '.format(U[0], U[1], U[2], U[3], RMSE, huber.cost)
textstr+= '{:8.6f} {:4d} {:4d}'.format(cost/LC_K3.shape[1], LC_K3.shape[1], LC_K.shape[1])


print textstr

